src/hotspot/cpu/x86/stubGenerator_x86_64_aes.cpp

```diff
@@ -3177,4 +3290,493 @@ void StubGenerator::aesgcm_encrypt(Register in, Register len, Register ct, Regis
+  for (int rnum = 1; rnum <= 8; rnum++) {
+    __ vpshufb(as_XMMRegister(rnum), as_XMMRegister(rnum), t5, Assembler::AVX_128bit); //perform a 16Byte swap
+  }
+
+  load_key(t_key, key, 16 * 0, rbx /*rscratch*/);
+  for (int rnum = 1; rnum <= 8; rnum++) {
+    __ vpxor(as_XMMRegister(rnum), as_XMMRegister(rnum), t_key, Assembler::AVX_128bit);
+  }
+
+  for (int i = 1; i <= 9; i++) {
+    load_key(t_key, key, 16 * i, rbx /*rscratch*/);
+    aesenc_step_avx2(t_key);
+  }
+
+  load_key(t_key, key, 16 * 10, rbx /*rscratch*/);
+  __ cmpl(rounds, 52);
+  __ jcc(Assembler::less, last_aes_rnd);
+
+  __ bind(aes_192);
+  aesenc_step_avx2(t_key);
+  load_key(t_key, key, 16 * 11, rbx /*rscratch*/);
+  aesenc_step_avx2(t_key);
+  load_key(t_key, key, 16 * 12, rbx /*rscratch*/);
+  __ cmpl(rounds, 60);
+  __ jcc(Assembler::less, last_aes_rnd);
+
+  __ bind(aes_256);
+  aesenc_step_avx2(t_key);
+  load_key(t_key, key, 16 * 13, rbx /*rscratch*/);
+  aesenc_step_avx2(t_key);
+  load_key(t_key, key, 16 * 14, rbx /*rscratch*/);
+
+  __ bind(last_aes_rnd);
+  for (int rnum = 1; rnum <= 8; rnum++) {
+    __ aesenclast(as_XMMRegister(rnum), t_key);
+  }
+
+  //The hash should end up in T3
+  __ movdqu(t3, t2);
+  // XOR and store data
+  for (int i = 0; i <= 7; i++) {
+    __ movdqu(t1, Address(in, pos, Address::times_1, 16 * i));
+    __ vpxor(as_XMMRegister(i + 1), as_XMMRegister(i + 1), t1, Assembler::AVX_128bit);
+    __ movdqu(Address(out, pos, Address::times_1, 16 * i), as_XMMRegister(i + 1));
+  }
+
+  __ cmpptr(ct, out);
+  __ jcc(Assembler::equal, skip_reload);
+  for (int i = 0; i <= 7; i++) {
+    __ movdqu(as_XMMRegister(i + 1), Address(in, pos, Address::times_1, 16 * i));
+  }
+
+  __ bind(skip_reload);
+  //Update len (r14) with the number of blocks processed
+  __ subl(len, 128);
+  __ addl(pos, 128);
+
+  for (int rnum = 1; rnum <= 8; rnum++) {
+    __ vpshufb(as_XMMRegister(rnum), as_XMMRegister(rnum), ExternalAddress(counter_shuffle_mask_addr()), Assembler::AVX_128bit, rbx /*rscratch*/);
+  }
+  // Combine GHASHed value with the corresponding ciphertext
+  __ vpxor(xmm1, xmm1, t3, Assembler::AVX_128bit);
+}
+
+void StubGenerator::aesgcm_avx2(Register in, Register len, Register ct, Register out, Register key,
+                                Register state, Register subkeyHtbl, Register counter) {
+  const Register pos = rax;
+  const Register rounds = r10;
+  const XMMRegister ctr_blockx = xmm9;
+  const XMMRegister aad_hashx = xmm8;
+  Label encrypt_done, encrypt_by_8_parallel, encrypt_by_8_new, encrypt_by_8, hash_last_8, enc_dec_done, generate_htbl_8_blks;
+
+  //This routine should be called only for message sizes of 128 bytes or more.
+  //Macro flow:
+  //process 8 16 byte blocks in initial_num_blocks.
+  //process 8 16 byte blocks at a time until all are done 'encrypt_by_8_new  followed by ghash_last_8'
+  __ xorl(pos, pos);
+
+  // Generate 8 constants for htbl
+  __ call(generate_htbl_8_blks, relocInfo::none);
```

Comment:
why didn't you inline `generateHtbl_8_block_avx2` here? This method is only used here as far as I can tell.